src/test/test-bitfield.c

```diff
@@ -0,0 +1,230 @@
+        TEST_BITS_SET(u64, 0, 1, 3, 4, 5, 8, 9, 10, 12, 14, 17, 18, 21, 23, 24, 25, 26, 32, 34, 39, 40, 42, 43, 45, 49, 50, 51, 56, 59, 61, 62);
+        TEST_BITS_CLEAR(u64, 2, 6, 7, 11, 13, 15, 16, 19, 20, 22, 27, 28, 29, 30, 31, 33, 35, 36, 37, 38, 41, 44, 46, 47, 48, 52, 53, 54, 55, 57, 58, 60, 63);
+        SET_BITS(u64, 63, 62, 7, 13, 38, 40);
+        TEST_BITS_SET(u64, 0, 1, 3, 4, 5, 7, 8, 9, 10, 12, 13, 14, 17, 18, 21, 23, 24, 25, 26, 32, 34, 38, 39, 40, 42, 43, 45, 49, 50, 51, 56, 59, 61, 62, 63);
+        TEST_BITS_CLEAR(u64, 2, 6, 11, 15, 16, 19, 20, 22, 27, 28, 29, 30, 31, 33, 35, 36, 37, 41, 44, 46, 47, 48, 52, 53, 54, 55, 57, 58, 60);
+        CLEAR_BIT(u64, 32);
+        TEST_BITS_SET(u64, 0, 1, 3, 4, 5, 7, 8, 9, 10, 12, 13, 14, 17, 18, 21, 23, 24, 25, 26, 34, 38, 39, 40, 42, 43, 45, 49, 50, 51, 56, 59, 61, 62, 63);
+        TEST_BITS_CLEAR(u64, 2, 6, 11, 15, 16, 19, 20, 22, 27, 28, 29, 30, 31, 32, 33, 35, 36, 37, 41, 44, 46, 47, 48, 52, 53, 54, 55, 57, 58, 60);
+        CLEAR_BITS(u64, 0, 2, 11, 63, 32, 58);
+        TEST_BITS_SET(u64, 1, 3, 4, 5, 7, 8, 9, 10, 12, 13, 14, 17, 18, 21, 23, 24, 25, 26, 34, 38, 39, 40, 42, 43, 45, 49, 50, 51, 56, 59, 61, 62);
+        TEST_BITS_CLEAR(u64, 0, 2, 6, 11, 15, 16, 19, 20, 22, 27, 28, 29, 30, 31, 32, 33, 35, 36, 37, 41, 44, 46, 47, 48, 52, 53, 54, 55, 57, 58, 60, 63);
+        expected64 = 0;
+        BIT_FOREACH(i, u64)
+                expected64 |= UINT64_C(1) << i;
+        assert_se(expected64 == u64);
+        u64 = 0;
+        TEST_BITS_CLEAR(u64, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63);
+        BIT_FOREACH(i, u64)
+                assert_se(0);
+        u64 = ~u64;
+        TEST_BITS_SET(u64, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63);
+        count = 0;
+        BIT_FOREACH(i, u64)
+                count++;
+        assert_se(count == 64);
+        uint64_t _u64 = u64;
+        SET_BITS(u64);
+        assert_se(_u64 == u64);
+        CLEAR_BITS(u64);
+        assert_se(_u64 == u64);
+}
+
+TEST_RET(bits_constant) {
+        assert_se(__builtin_constant_p(INDEX_TO_MASK(uint8_t, 1)));
+        assert_se(__builtin_constant_p(INDEX_TO_MASK(uint16_t, 1)));
+        assert_se(__builtin_constant_p(INDEX_TO_MASK(uint32_t, 1)));
+        assert_se(__builtin_constant_p(INDEX_TO_MASK(uint64_t, 1)));
+
+        assert_se(__builtin_constant_p(BIT_SET((uint8_t)2, 1)));
+        assert_se(__builtin_constant_p(BIT_SET((uint16_t)2, 1)));
+        assert_se(__builtin_constant_p(BIT_SET((uint32_t)2, 1)));
+        assert_se(__builtin_constant_p(BIT_SET((uint64_t)2, 1)));
```

Comment:
btw, would make sense to make these assert_cc(), no?

(and move them outside of any function then)

Reply:
ack, though it has to be in a function, as the compiler complains if ({}) is used outside a function. Of course technically since it's inside assert_cc() it doesn't actually produce any code but I guess it's the preprocessor complaining.

Reply:
also i tried assert_cc() before and the clang --optimization=0 test build was failing, but i think i was testing it differently then, the assert_cc() with __builtin_constant_p() seems to work ok